In [11]:
import pandas as pd
from sqlalchemy import create_engine
import psycopg2
from psycopg2.extras import RealDictCursor

#import matplotlib.pyplot as plt
import numpy as np
#import os
#from pathlib import Path

In [2]:
uri = "postgresql+psycopg2://avnadmin:vr74tpu3569e0um4@pg-260d0c55-tatwan77-38eb.aivencloud.com:20669/defaultdb?sslmode=require"

In [22]:
print(np.__version__)
print(pd.__version__)
!python --version

1.20.2
0.24.2
Python 3.7.10


In [24]:
!conda list

# packages in environment at C:\Users\chakravartiraghavan\anaconda3\envs\sqlenv:
#
# Name                    Version                   Build  Channel
_anaconda_depends         2020.07                  py37_0  
alabaster                 0.7.12                   py37_0  
anaconda                  custom                   py37_1  
anaconda-client           1.7.2                    py37_0  
anaconda-project          0.10.0             pyhd3eb1b0_0  
anyio                     2.2.0            py37haa95532_2  
appdirs                   1.4.4                      py_0  
argh                      0.26.2                   py37_0  
argon2-cffi               20.1.0           py37h2bbff1b_1  
asn1crypto                1.4.0                      py_0  
astroid                   2.5              py37haa95532_1  
astropy                   4.2.1            py37h2bbff1b_1  
async_generator           1.10             py37h28b3542_0  
atomicwrites              1.4.0                      py_0  
attrs     

In [13]:
engine = create_engine(uri)

In [14]:
# test connection
tmp1 = '''
select
	m.name
from
	merchant m
'''
df = pd.read_sql(tmp1, engine)
df.head()

name
0      Murphy, Heath and Fields
1                   Riggs-Adams
2  Sanders, Parks and Mcfarland
3                Mccarty-Thomas
4                Miller-Blevins

In [15]:
# get card holder time and amount
info = '''
  select t.card, t.amount, ch.name, t.date
  from
    transaction t
  inner join credit_card as c on t.card = c.card
  inner join card_holder as ch on c.cardholder_id = ch.id
  
  '''
info_transactions_df = pd.read_sql(info, engine)
print(info_transactions_df.head())
print(len(info_transactions_df))

               card  amount               name                date
0  3561954487988605    6.22        John Martin 2018-01-01 21:35:10
1  5135837688671496    3.83        John Martin 2018-01-01 21:43:12
2   213193946980303    9.61  Matthew Gutierrez 2018-01-01 22:41:21
3  4263694062533017   19.03     Danielle Green 2018-01-01 23:13:30
4     4498002758300    2.95      Malik Carlson 2018-01-01 23:15:10
3500


In [16]:
info_transactions_df.sort_values(by=['amount'], inplace=True)
print(info_transactions_df.head())
print('total transactions = ', len(info_transactions_df))

                  card  amount           name                date
341   3561954487988605    0.51    John Martin 2018-02-04 19:02:09
1949  4506405265172173    0.52  Kevin Spencer 2018-07-17 22:11:50
1918  4506405265172173    0.53  Kevin Spencer 2018-07-14 17:44:09
1023  5500708021555307    0.54  Crystal Clark 2018-04-13 04:50:37
1070  5500708021555307    0.56  Crystal Clark 2018-04-18 06:51:00
total transactions =  3500


In [17]:
#df.loc[df['column_name'].isin(some_values)]
#small_trasactions_df = info_transactions_df.loc[info_transactions_df['amount'].isin(info_transactions_df['amount']<2)]
#small_trasactions_df.head()
#info_transactions_df.loc[info_transactions_df['amount'].isin(info_transactions_df['amount']<2)]
less_than2_df = info_transactions_df['amount']<2
less_than2_df = info_transactions_df[less_than2_df]
print(less_than2_df.head())
print('Less than $2 transactions = ', len(less_than2_df))

                  card  amount           name                date
341   3561954487988605    0.51    John Martin 2018-02-04 19:02:09
1949  4506405265172173    0.52  Kevin Spencer 2018-07-17 22:11:50
1918  4506405265172173    0.53  Kevin Spencer 2018-07-14 17:44:09
1023  5500708021555307    0.54  Crystal Clark 2018-04-13 04:50:37
1070  5500708021555307    0.56  Crystal Clark 2018-04-18 06:51:00
Less than $2 transactions =  350


### We see that the Total transactions is 3500 and less than $2 are 350. Which is 10%

In [18]:
type(info_transactions_df)
#info_transactions_df.info(verbose=True)
#info_transactions_df.plot()
info_transactions_df.info(verbose=True)

TypeError: Cannot interpret '<attribute 'dtype' of 'numpy.generic' objects>' as a data type

In [25]:
print(len(small_trasactions_df))

1


In [8]:
# get transaction less than $2
tr1 = '''
  select t.card, count(*) 
  from
    transaction t
  where t.amount <= 2
  group by t.card
  '''
small_transactions_df = pd.read_sql(tr1, engine)
print(small_transactions_df.head())
print(len(small_transactions_df))

               card  count
0  4027907156459098      6
1      501809222273      7
2  5500708021555307      8
3  3561072557118696      8
4    30078299053512      3
53


In [9]:
# get transaction > $2
tr_big = '''
  select t.card, count(*) 
  from
    transaction t
  where t.amount > 2
  group by t.card
  '''
bigger_transactions_df = pd.read_sql(tr_big, engine)
print(bigger_transactions_df.head())
print(len(bigger_transactions_df))

               card  count
0  4027907156459098     66
1      501809222273     60
2  5500708021555307     63
3  3561072557118696     56
4    30078299053512     55
53


####  we see that roughly 10% of transactions are less than $2 - which may not be fraud especially since it seems to be spread out.

In [74]:
# transactions getting data between 7 and 9 am.
transactions_7_9_cnt = '''
select
   t.card, count(*)
from
	transaction t
where cast(t.date as time) between cast('07:00:00' as time) and cast('09:00:00' as time)
group by card
'''
transactions_7_9 = '''
select
   *
from
	transaction t
where cast(t.date as time) between cast('07:00:00' as time) and cast('09:00:00' as time)
'''
# where date > '2018-12-01'
# 	cast(t.date as time) as tm, date
trn_7_9_cnt_df = pd.read_sql(transactions_7_9_cnt, engine)
print(trn_7_9_cnt_df.head())
trn_7_9_df = pd.read_sql(transactions_7_9, engine)
print(trn_7_9_df.head())

              card  count
0  180098539019105      7
1  213193946980303      3
2   30063281385429      4
3   30078299053512      5
4   30142966699187      4
     id                date  amount              card  id_merchant
0  2387 2018-01-03 08:44:46    3.89  4027907156459098           74
1  2077 2018-01-05 07:19:27    1.36   344119623920892           30
2  2890 2018-01-06 07:00:32   18.35  3561954487988605           76
3  3291 2018-01-06 08:42:50   18.72  3535651398328201           76
4  2557 2018-01-06 08:49:34   16.28  4268491956169254          101


In [75]:
trn_7_9_cnt_df.rename(columns={"count": "7_9_count"}, inplace=True)

In [76]:
trn_7_9_cnt_df = trn_7_9_cnt_df.set_index(['card'])
print(trn_7_9_cnt_df.head())

                 7_9_count
card                      
180098539019105          7
213193946980303          3
30063281385429           4
30078299053512           5
30142966699187           4


In [77]:
len(trn_7_9_cnt_df)

52

In [78]:
# transactions outside between 7 and 9 am.
transactions_rest_cnt = '''
select
   t.card, count(*)
from
	transaction t
where cast(t.date as time) not between cast('07:00:00' as time) and cast('09:00:00' as time)
group  by card
'''
transactions_rest = '''
select
   *
from
	transaction t
where cast(t.date as time) not between cast('07:00:00' as time) and cast('09:00:00' as time)
'''
# where date > '2018-12-01'
# 	cast(t.date as time) as tm, date
trn1_df_cnt = pd.read_sql(transactions_rest_cnt, engine)
print(trn1_df_cnt.head())
trn1_df = pd.read_sql(transactions_rest, engine)
print(trn1_df.head())

               card  count
0  4027907156459098     65
1      501809222273     64
2  5500708021555307     66
3  3561072557118696     58
4    30078299053512     53
     id                date  amount              card  id_merchant
0   222 2018-01-01 21:35:10    6.22  3561954487988605           69
1  2045 2018-01-01 21:43:12    3.83  5135837688671496           85
2   395 2018-01-01 22:41:21    9.61   213193946980303           82
3  3309 2018-01-01 23:13:30   19.03  4263694062533017            5
4   567 2018-01-01 23:15:10    2.95     4498002758300           64


In [79]:
trn1_df_cnt.rename(columns={"count": "rest_count"}, inplace=True)

In [80]:
len(trn1_df_cnt)

53

In [81]:
trn1_df_cnt = trn1_df_cnt.set_index(['card'])
print(trn1_df.head())

     id                date  amount              card  id_merchant
0   222 2018-01-01 21:35:10    6.22  3561954487988605           69
1  2045 2018-01-01 21:43:12    3.83  5135837688671496           85
2   395 2018-01-01 22:41:21    9.61   213193946980303           82
3  3309 2018-01-01 23:13:30   19.03  4263694062533017            5
4   567 2018-01-01 23:15:10    2.95     4498002758300           64


In [82]:
len(trn1_df_cnt)

53

In [83]:
combined_df = pd.concat([trn_7_9_cnt_df, trn1_df_cnt], axis='columns', join='inner')
print(combined_df.head())

                 7_9_count  rest_count
card                                  
180098539019105          7          66
213193946980303          3          67
30063281385429           4          54
30078299053512           5          53
30142966699187           4          53


In [84]:
perc_combined_df = combined_df['7_9_count'] / combined_df['rest_count']

In [106]:
# check if more than 15% are within 7AM to 9AM
for i in perc_combined_df:
    if i > .15:
        print(i)

0.15555555555555556
0.15714285714285714
0.16326530612244897
0.18181818181818182
0.15384615384615385
0.16666666666666666


### Based no the above data there is no anamoly for between 7AM to 9AM